src/ProxyClient/Request/RequestQueue.php

```diff
@@ -0,0 +1,221 @@
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+class RequestQueue implements \Countable
+{
+    /**
+     * HTTP cache servers
+     *
+     * @var UriInterface[]
+     */
+    private $servers = [];
+
+    /**
+     * Application base URI
+     *
+     * @var UriInterface | null
+     */
+    private $baseUri;
+
+    /**
+     * @var InvalidationRequest[]
+     */
+    private $queue = [];
+
+    /**
+     * Constructor
+     *
+     * @param array $servers
+     * @param null  $baseUri
+     */
+    public function __construct(
+        array $servers,
+        $baseUri = null
+    ) {
+        $this->setServers($servers);
+        $this->setBaseUri($baseUri);
+    }
+
+    /**
+     * Add invalidation request
+     *
+     * @param InvalidationRequest $request
+     */
+    public function add(InvalidationRequest $request)
+    {
+        $signature = $request->getSignature();
+        
+        if (!isset($this->queue[$signature])) {
+            $this->queue[$signature] = $request;
+        }
+    }
+
+    /**
+     * Clear request queue
+     */
+    public function clear()
+    {
+        $this->queue = [];
+    }
+    
+    public function count()
+    {
+        return count($this->queue);
+    }
+
+    /**
+     * Get each invalidation request replicated over all HTTP caching servers
+     *
+     * @return RequestInterface[]
+     */
+    public function all()
+    {
+        $requests = [];
+        foreach ($this->queue as $request) {
+            $uri = $request->getUri();
+            
+            // If a base URI is configured, try to make partial invalidation
+            // requests complete.
+            if ($this->baseUri) {
+                if ($uri->getHost()) {
+                    // Absolute URI: does it already have a scheme?
+                    if (!$uri->getScheme() && $this->baseUri->getScheme() !== '') {
+                        $uri = $uri->withScheme($this->baseUri->getScheme());
+                    }
+                } else {
+                    // Relative URI
+                    if ($this->baseUri->getHost() !== '') {
+                        $uri = $uri->withHost($this->baseUri->getHost());
+                    }
+                    
+                    if ($this->baseUri->getPort()) {
+                        $uri = $uri->withPort($this->baseUri->getPort());
+                    }
+                    
+                    // Base path
+                    if ($this->baseUri->getPath() !== '') {
+                        $path = $this->baseUri->getPath() . '/' . ltrim($uri->getPath(), '/');
+                        $uri = $uri->withPath($path);
+                    }
+                }
+            }
+            
+            // Close connections to make sure invalidation (PURGE/BAN) requests
+            // will not interfere with content (GET) requests.
+            $request = $request->withUri($uri)->withHeader('Connection', 'Close');
```

Comment:
Sending `Connection: Close` only for the last invalidation request makes the timing issue come back, so as far as I can see we really need to close *each* invalidation request’s connection. Doing this from the client-side won’t work either, because that requires cURL/Guzzle-specifics, and we’re trying to get rid of those. 

All in all, I guess the performance impact won’t be so big: unless your sending invalidating many URLs per second, which defeats the purpose of caching those URLs anyway.